Project activation

jennybc opened the pull request:
Fixes #390 When rstudio = TRUE, don't change active project in create_package()

Generally tightens up project activation:

  * ~Quietly attempt to activate project on load (should this be on attach?)~
  * Announce project activation, by default
  * Rename `is_proj(path)` to `possibly_in_proj(path)` to better reflect what it actually reports
  * ~Store current project *and the previous*, so we can have `proj_revert()`~

hadley commented on R/zzz.R:
```diff
@@ -0,0 +1,3 @@
+.onLoad <- function(libname, pkgname) {
+  try(proj_set(".", quiet = TRUE), silent = TRUE)
+}
```
Why is this needed now, but not previously? I don't really understand how project activation is supposed to work.

jennybc replied:
Oh, there is no new need! In most usage scenarios, this doesn't matter. You do notice it when developing usethis itself. This activates a project after `load_all()`, which is just nice.

I would say project activation is supposed to work "on first demand".

Another consequence of auto-activate, more relevant to other people: All `use_*()` functions that operate on a project eventually make a `proj_get()` and that will activate a project. But if the first usage is of a helper like `write_union()`, that won't activate a project, but it *will* honour an already-activated project. This does not change what is written or where, but makes the path reported in the message nicer (i.e. relative to the project instead of absolute).

jennybc replied:
If auto-activation seems like a bad idea, I could also deal with the usethis development hangnail via something in the test helper file.

jennybc commented:
BTW here's what the new messaging/flow looks like. Project activation is announced, but is no longer in bold and red. If new project was launched in RStudio, host process reverts to original state.

<img width="851" alt="screen shot 2018-06-27 at 7 26 17 am" src="https://user-images.githubusercontent.com/599454/41980581-aab0a6c8-79db-11e8-91a2-7c255651bde2.png">

hadley commented:
Candidate function (name needs work):

```R
open_rstudio <- function(open, old_project) {
  if (!open)
    return()
  
  fresh_rstudio <- open_project(proj_get())
  if (!fresh_rstudio) 
    return()
  
  proj_set(old_project, force = TRUE)
}
```

(Assuming we revert `proj_revert()` and use the pattern in the previous PR)

jennybc commented:
Also record the unresolved dilemmas about the relationship between `write_*()`-type helpers and the active project.

jennybc commented on R/create.R:
```diff
@@ -194,30 +193,36 @@ create_from_github <- function(repo_spec,
-  rproj_path <- rproj_path(path)
+## `rstudio` arg here is about whether to attempt a launch in RStudio
+## `rstudio` arg of `create_*()` functions is about whether to add .Rproj file
+open_project <- function(path, restore = NA, rstudio = NA) {
```
The default for `restore` must be `NA`, because `NULL` is a valid value to restore, i.e. we could need to restore the state of "no active project".

jennybc commented on tests/testthat/helper.R:
```diff
@@ -1,3 +1,6 @@
+## attempt to activate a project, which is nice during development
+try(proj_set("."))
+
```
New solution to the "no active project" problem during usethis development.

jennybc commented:
@hadley I think is good to go.

  * Added principles.md to document design decisions and dilemmas.
  * Back to old way of restoring project state, i.e. reverted `proj_revert()`.
  * Refactored `open_project()` to do project restoration.

jennybc commented:
Spurious AppVeyor failure coming from fact that we depend on new styler release and Windows binaries aren't up yet.

hadley commented on R/create.R:
```diff
@@ -194,30 +193,36 @@ create_from_github <- function(repo_spec,
+      proj_set(restore, force = TRUE)
+    }
+    return(invisible(TRUE))
   }
```
(Only commenting because we've been discussing early returns a lot)

This doesn't feel quite right as an early return to me. It would feel better as true and false branches of an if statement, because it's more symmetric: either open in RStudio or change the working directory.

hadley commented on principles.md:
```diff
@@ -0,0 +1,31 @@
+
+### Activation upon load or attach? No.
+
+We've contemplated project activation in `.onLoad()` or `.onAttach()`, but it's not clear which is more appropriate. Which suggests that neither is appropriate. If we ever do this, `zzz.R` would include something like this:
```
Two more reasons not to activate on load/attach:

* If on attach, would mean that an package that imported usethis would also need to set on attach

* Current workflow allows you to open R, then `setwd()`, then usethis (not a workflow that we recommend, but one that is common in practice)

hadley commented on principles.md:
```diff
@@ -0,0 +1,31 @@
+
+## Active project
+
+Many usethis functions act on the **active project**, the path to which is stored in the internal environment `proj`, specifically in `proj$cur`. We do this instead of constantly passing around a base path. It is implied that such functions create or modify files inside the active project. This is mostly true of `use_*()` functions, though there are exceptions. For example, `use_course()` makes n ...
```
We do this instead of constantly passing around a base path or relying on the working directory.
